### PR TITLE
add tikv metrics proxy

### PR DESCRIPTION
**Pull request by LinuxGit**
TiKV metrics proxy is a proxy for the TiKV server's prometheus metrics. The `tikv_metrics_proxy` exposes metrics via HTTP on the `/metrics` endpoint.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,17 +1,21 @@
+LDFLAGS += -X "github.com/pingcap/tidb-inspect-tools/pkg/utils.Version=1.0.0"
```
Comment (GregoryIan) [inline Makefile]: `1.0.0`?

File: README.md [hunk 2]
```diff
@@ -25,5 +25,9 @@ tidb-inspect-tools are some useful inspection tool collections for TiDB.
 
   A tool for checking the TiDB server's health by sending SQL queries to it. The `tidb_exporter` exposes metrics via HTTP on the `/metrics` endpoint for Prometheus.
 
+- [tikv_metrics_proxy](https://github.com/pingcap/tidb-inspect-tools/tree/master/tikv_metrics_proxy)
+
+  A proxy for the TiKV server's prometheus metrics.
```
Comment (GregoryIan) [inline README.md]: to expose Prometheus metrics of TiKV server

File: tikv_metrics_proxy/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+```
+
+#### Running
+
+```
+$ ./tikv_metrics_proxy -h
+usage: tikv_metrics_proxy [<flags>]
+
+Flags:
+  -h, --help              Show context-sensitive help (also try --help-long and --help-man).
```
Comment (GregoryIan) [inline tikv_metrics_proxy/README.md]: What are `help-long` and `help-man`?

File: tikv_metrics_proxy/main.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+var (
+	stores = []string{}
+)
+
+type tikvOpts struct {
+	addrs string
+}
+
+// ParseHostPortAddr returns a host:port list
+func ParseHostPortAddr(s string) ([]string, error) {
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: We have written this function earily, can u reuse it?

File: tikv_metrics_proxy/main.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+		listenAddress = kingpin.Flag("web.listen-address", "Address on which to expose metrics and web interface.").Default(":9600").String()
+		metricsPath   = kingpin.Flag("web.telemetry-path", "Path under which to expose metrics.").Default("/metrics").String()
+		logFile       = kingpin.Flag("log-file", "Log file path.").Default("").String()
+		logLevel      = kingpin.Flag("log-level", "Log level: debug, info, warn, error, fatal.").Default("info").String()
+		logRotate     = kingpin.Flag("log-rotate", "Log file rotate type: hour/day.").Default("day").String()
+
+		opts = tikvOpts{}
+	)
+
+	kingpin.Flag("tikv.addrs", "Addresses (host:port) of TiKV instances, comma separated.").Default("").StringVar(&opts.addrs)
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: Unify it with codes above?

File: tikv_metrics_proxy/main.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+		mutex.Lock()
+		for _, m := range metricFamilies {
+			allMetrics = append(allMetrics, m)
+		}
+		mutex.Unlock()
+	}
+
+	for _, store := range stores {
+		wg.Add(1)
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: No need to call `go` and `wait` to control signle process

File: tikv_metrics_proxy/main.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+func getMetricFamilies() []*dto.MetricFamily {
+	wg := sync.WaitGroup{}
+	mutex := &sync.Mutex{}
+	ctx := context.Background()
+	allMetrics := make([]*dto.MetricFamily, 0, 1024)
+
+	getStoreMetrics := func(store string) {
+		defer wg.Done()
+
+		tikvConn, err := grpc.Dial(store, grpc.WithInsecure())
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: Only need to initial it once?

File: tikv_metrics_proxy/main.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+		metricFamilies, err := parser.TextToMetricFamilies(bytes.NewBufferString(mData))
+		if err != nil {
+			log.Errorf("store '%s', TextToMetricFamilies error, %v", store, err)
+			return
+		}
+
+		sanitizeLabels(metricFamilies, labels)
+
+		mutex.Lock()
+		for _, m := range metricFamilies {
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: No need lock

File: tikv_metrics_proxy/README.md (hunk 3 above)
Comment (LinuxGit) [reply tikv_metrics_proxy/README.md]: It's build-in help message in kingpin package:
https://github.com/alecthomas/kingpin/blob/7e2d1284f714168af95785926b6c7da2aa0abd21/app.go#L62

**Comment by LinuxGit** [comment]
It supports multiple tikv servers, so need wait group and lock.

File: tikv_metrics_proxy/main.go (hunk 4 above)
Comment (LinuxGit) [reply tikv_metrics_proxy/main.go]: I've moved it to pkg utils.

File: tikv_metrics_proxy/main.go (hunk 5 above)
Comment (LinuxGit) [reply tikv_metrics_proxy/main.go]: I'm going to add tls flags in tikvOpts.

**Comment by LinuxGit** [comment]
@tennix yes. I'm going to add vendor in another pr or after reviewing.

File: tikv_metrics_proxy/client.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+			return nil, err
+		}
+	}
+	return conn, nil
+}
+
+func (c *rpcClient) createConn(addr string) (*grpc.ClientConn, error) {
+	c.Lock()
+	defer c.Unlock()
+	conn, ok := c.conns[addr]
```
Comment (GregoryIan) [inline tikv_metrics_proxy/client.go]: create just does creation

File: tikv_metrics_proxy/client.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+func (c *rpcClient) closeConns() {
+	c.Lock()
+	if !c.isClosed {
+		c.isClosed = true
+		// close all connections
+		for _, conn := range c.conns {
+			conn.Close()
+		}
+	}
```
Comment (GregoryIan) [inline tikv_metrics_proxy/client.go]: reset `c.conns`?

File: tikv_metrics_proxy/README.md [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+      --web.listen-address=":9600"
+                          Address on which to expose metrics and web interface.
+      --web.telemetry-path="/metrics"
```
Comment (siddontang) [inline tikv_metrics_proxy/README.md]: `/metrics` is the common path for Prometheus, I think we don't need to provide config for it

File: tikv_metrics_proxy/client.go [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"sync"
+
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"google.golang.org/grpc"
+)
+
+func newConn(addr string) (*grpc.ClientConn, error) {
+	conn, err := grpc.Dial(addr, grpc.WithInsecure())
```
[the remark this answers is not in the document]
Comment (LinuxGit) [reply tikv_metrics_proxy/client.go]: Yes. I'll add tls support in this pr. I prepare to add it in next pr originally.

File: tikv_metrics_proxy/README.md (hunk 11 above)
Comment (LinuxGit) [reply tikv_metrics_proxy/README.md]: Yes. Providing `web.telemetry-path` flag is common way for prometheus exporters, like node_exporter and influxdb_exporter. We could hide it in ansible or docker deployment.

File: tikv_metrics_proxy/main.go [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+			}
+			sort.Sort(prometheus.LabelPairSorter(m.Label))
+		}
+	}
+}
+
+func (e *Exporter) getMetricFamilies() []*dto.MetricFamily {
+	wg := sync.WaitGroup{}
+	mutex := &sync.Mutex{}
+	ctx := context.Background()
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: do u want to control this function by `context`?

File: tikv_metrics_proxy/main.go [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+
+	getStoreMetrics := func(store string) {
+		defer wg.Done()
+
+		tikvConn, err := e.client.getConn(store)
+		if err != nil {
+			return
+		}
+
+		tikvClient := debugpb.NewDebugClient(tikvConn)
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: init it once?

File: tikv_metrics_proxy/main.go [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+		metrics, err := tikvClient.GetMetrics(ctx, &debugpb.GetMetricsRequest{})
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: use timeout context?

File: tikv_metrics_proxy/client.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (LinuxGit) [reply tikv_metrics_proxy/client.go]: I've added tls support in this pr, refer to tidb's code.

File: tikv_metrics_proxy/client.go [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+	}
+	c.conns[addr] = conn
+	return conn, nil
+}
+
+func (c *rpcClient) closeConns() {
+	c.Lock()
+	if !c.isClosed {
+		c.isClosed = true
+		for i, conn := range c.conns {
```
Comment (GregoryIan) [inline tikv_metrics_proxy/client.go]: `c.conns = nil`

File: tikv_metrics_proxy/client.go [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+		if err != nil {
+			return nil, err
+		}
+	}
+	return conn, nil
+}
+
+func (c *rpcClient) createConn(addr string) (*grpc.ClientConn, error) {
+	c.Lock()
+	defer c.Unlock()
```
Comment (GregoryIan) [inline tikv_metrics_proxy/client.go]: also check `c.isClosed`

File: tikv_metrics_proxy/main.go [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+	go func() {
+		sig := <-sc
+		log.Infof("got signal [%d] to exit", sig)
+
+		exporter.client.closeConns()
+		os.Exit(0)
+	}()
+
+	log.Info("listening on", *listenAddress)
+	log.Fatal(http.ListenAndServe(*listenAddress, nil))
```
Comment (GregoryIan) [inline tikv_metrics_proxy/main.go]: close connections？

**Comment by GregoryIan** [comment]
Rest LGTM

File: tikv_metrics_proxy/main.go [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+		if err != nil {
+			return
+		}
+
+		tikvClient := debugpb.NewDebugClient(tikvConn)
+		ctx, cancel := context.WithTimeout(context.Background(), getMetricsTimeout)
+		defer cancel()
+		metrics, err := tikvClient.GetMetrics(ctx, &debugpb.GetMetricsRequest{})
+		if err != nil {
+			log.Errorf("tikv store '%s', get metrics error, %v", store, err)
```
Comment (siddontang) [inline tikv_metrics_proxy/main.go]: do we need to remove the connection if meet any error?

Comment (LinuxGit) [reply tikv_metrics_proxy/main.go]: grpc client has reconnection mechanism，and DefaultBackoffMaxDelay is 120s, I add `grpc.WithBackoffMaxDelay(time.Second*5)` option in the new commit.

**Pull request merged**